source/common/buffer/buffer_impl.cc

```diff
@@ -41,6 +41,19 @@ void OwnedImpl::add(const Instance& data) {
   }
 }
 
+void OwnedImpl::prepend(const void* data, uint64_t size) {
+  evbuffer_prepend(buffer_.get(), data, size);
+}
+
+void OwnedImpl::prepend(const std::string& data) {
+  evbuffer_prepend(buffer_.get(), data.data(), data.size());
+}
+
+void OwnedImpl::prepend(Instance& data) {
+  evbuffer_prepend_buffer(buffer_.get(), static_cast<LibEventInstance&>(data).buffer().get());
+  data.drain(data.length());
+}
```

Comment:
I believe this is going to need a call to postProcess just as the other calls with static_cast<LibEventInstance> do

Can you add tests for all three prepend functions to watermark buffer, to make sure that no matter how we prepend, if we go above the high watermark we trigger the correct callbacks?

Reply:
I've added the call to `postProcess`. Also, I've added the prepend functions to WatermarkBuffer along with test cases to ensure watermarks are triggered appropriately.